frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1446 @@
+			assert!(task.fully_complies_with(chain_limits));
+
+			Self::deposit_event(Event::<T>::Migrated(
+				task.dyn_top_items,
+				task.dyn_child_items,
+				MigrationCompute::Unsigned,
+			));
+			MigrationProcess::<T>::put(task);
+
+			Ok(().into())
+		}
+
+		/// Continue the migration for the given `limits`.
+		///
+		/// The dispatch origin of this call can be any signed account.
+		///
+		/// This transaction has NO MONETARY INCENTIVES. calling it will only incur transaction fees
+		/// on the caller, with no rewards paid out.
+		///
+		/// The sum of the byte length of all the data read must be provided for up-front
+		/// fee-payment and weighing.
+		#[pallet::weight(
+			// the migration process
+			Pallet::<T>::dynamic_weight(limits.item, * real_size)
+			// rest of the operations, like deposit etc.
+			+ T::WeightInfo::continue_migrate()
+		)]
+		pub fn continue_migrate(
+			origin: OriginFor<T>,
+			limits: MigrationLimits,
+			real_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			let max_limits = T::SignedMigrationMaxLimits::get();
+			ensure!(
+				limits.size <= max_limits.size && limits.item <= max_limits.item,
+				"max signed limits not respected"
+			);
+
+			// ensure they can pay more than the fee.
+			let deposit = T::SignedDepositPerItem::get().saturating_mul(limits.item.into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut task = Self::migration_process();
+			task.migrate_until_exhaustion(limits);
+
+			// ensure that the migration witness data was correct.
+			if real_size != task.dyn_size {
+				// let the imbalance burn.
+				let (_imbalance, _remainder) = T::Currency::slash(&who, deposit);
+				debug_assert!(_remainder.is_zero());
+				return Err("Wrong witness data".into())
+			}
+
+			Self::deposit_event(Event::<T>::Migrated(
+				task.dyn_top_items,
+				task.dyn_child_items,
+				MigrationCompute::Signed,
+			));
+
+			MigrationProcess::<T>::put(task);
+			Ok(().into())
+		}
+
+		/// Migrate the list of top keys by iterating each of them one by one.
+		///
+		/// This does not affect the global migration process tracker ([`MigrationProcess`]), and
+		/// should only be used in case any keys are leftover due to a bug.
+		#[pallet::weight(Pallet::<T>::dynamic_weight(keys.len() as u32, *total_size))]
+		pub fn migrate_custom_top(
+			origin: OriginFor<T>,
+			keys: Vec<Vec<u8>>,
+			total_size: u32,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			// ensure they can pay more than the fee.
+			let deposit = T::SignedDepositPerItem::get().saturating_mul((keys.len() as u32).into());
+			ensure!(T::Currency::can_slash(&who, deposit), "not enough funds");
+
+			let mut dyn_size = 0u32;
+			for key in &keys {
+				if let Some(data) = sp_io::storage::get(&key) {
+					dyn_size = dyn_size.saturating_add(data.len() as u32);
+					sp_io::storage::set(key, &data);
+				}
+			}
+
+			if dyn_size != total_size {
```

Comment:
since the keys are _exactly_ known in this case, I think it is fair to ask the user to pass us exactly their size as well. Albeit what you are proposing is harmless.

Reply:
how would the key be exactly none ? the caller cannot know what extrinsics are included before his extrinsic.

Reply:
updated this one as well to support an overestimate. Refund not strictly needed as there  this is not a free transaction, and it is not intended to be used except for emergencies.

Reply:
https://github.com/paritytech/substrate/pull/10073/commits/1626d196a1b288862df62bda571e01f8e865e791